log/DESIGN.md

```diff
@@ -0,0 +1,287 @@
+# Logs Bridge API
+
+OpenTelemetry Logs tracking issue at [#4696](https://github.com/open-telemetry/opentelemetry-go/issues/4696).
+
+## Abstract
+
+We propose adding a `go.opentelemetry.io/otel/log` Go module which will provide
+[Logs Bridge API](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/).
+
+## Background
+
+They key challenge is to create a well-performant API compliant with the specification.
+Performance is seen as one of the most imporatant charactristics of logging libraries in Go.
+
+## Design
+
+This proposed design aims to:
+
+- be specification compliant,
+- have similar API to Trace and Metrics API,
+- take advantage of both OpenTelemetry and `slog` experience to achieve acceptable performance.
+
+### Module structure
+
+The Go module consits of the following packages:
+
+- `go.opentelemetry.io/otel/log`
+- `go.opentelemetry.io/otel/log/embedded`
+- `go.opentelemetry.io/otel/log/noop`
+
+### LoggerProvider
+
+The [`LoggerProvider` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#loggerprovider)
+is defined as an interface.
+
+```go
+type LoggerProvider interface{
+	embedded.LoggerProvider
+	Logger(name string, options ...LoggerOption) Logger
+}
+```
+
+### Logger
+
+The [`Logger` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#logger)
+is defined as an interface.
+
+```go
+type Logger interface{
+	embedded.Logger
+	Emit(ctx context.Context, options ...RecordOption)
```

Comment:
This allocates a slice of options for every Emit call (I think) whereas the version that accepts a Record doesn't. Do we consider this negligible?

This seems to be somewhat in contradiction with the design goal of performance, in comparison with the version that accepts a Record struct and which is rejected. 

It seems we are calling out performance as a key element of the design but don't seem to follow through on it. Should benchmarking be done _before_ the API choice is made to show performance is good enough?

Reply:
Makes sense 👍  I am leaning to even change the proposal.

Reply:
I am on-board with switching to Emit(Record), especially given this API isn't expected to be primarily a bridge API.

Reply:
My benchmark code just in case you want to see what it was about:
```go
func BenchmarkRecord(b *testing.B) {
	t := time.Now()
	for i := 0; i < b.N; i++ {
		emitRecord(&Record{Timestamp: t})
	}
}

//go:noinline
func emitRecord(r *Record) {

}

func BenchmarkRecordOption(b *testing.B) {
	t := time.Now()
	for i := 0; i < b.N; i++ {
		emitWithOptions(withTimestamp{t: t})
	}
}

func emitWithOptions(options ...RecordOption) {
	r := Record{}
	for i := 0; i < len(options); i++ {
		options[i].apply(&r)
	}

}

type RecordOption interface {
	apply(r *Record)
}

type withTimestamp struct {
	t time.Time
}

func (w withTimestamp) apply(r *Record) {
	r.Timestamp = w.t
}
```

Reply:
Just for reference. `slog` passes the records by value in `Handle`: https://pkg.go.dev/log/slog#Handler

I did a quick benchmark and it also seems faster.

Of course we should double-check it after the prototype is implemented with more realistic scenarios.

Results:

```
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
BenchmarkRecordValue-16         1000000000               1.144 ns/op           0 B/op          0 allocs/op
BenchmarkRecordPtr-16           749360420                1.594 ns/op           0 B/op          0 allocs/op
BenchmarkRecordOption-16        19574589                58.21 ns/op           48 B/op          2 allocs/op
```

Code:

```go
package main

import (
	"testing"
	"time"
)

func BenchmarkRecordValue(b *testing.B) {
	t := time.Now()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		emitRecordValue(Record{Timestamp: t})
	}
}

func BenchmarkRecordPtr(b *testing.B) {
	t := time.Now()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		emitRecordPtr(&Record{Timestamp: t})
	}
}

func BenchmarkRecordOption(b *testing.B) {
	t := time.Now()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		emitWithOptions(withTimestamp{t: t})
	}
}

type Record struct {
	Timestamp time.Time
}

//go:noinline
func emitRecordPtr(r *Record) {

}

//go:noinline
func emitRecordValue(r Record) {

}

func emitWithOptions(options ...RecordOption) {
	r := Record{}
	for i := 0; i < len(options); i++ {
		options[i].apply(&r)
	}

}

type RecordOption interface {
	apply(r *Record)
}

type withTimestamp struct {
	t time.Time
}

func (w withTimestamp) apply(r *Record) {
	r.Timestamp = w.t
}
```

Reply:
Nit: the `Record` struct in your benchmark is not representative of what will be in the final code, with just one Timestamp field. This may sqew the by-value vs by-pointer results. I used the `Record` struct with full set of fields proposed in this PR.

Reply:
I added more fields and changed the benchmark to a more realistic one

```go
func BenchmarkRecordValue(b *testing.B) {
	for i := 0; i < b.N; i++ {
		emitRecordValue(Record{Timestamp: t, Severity: lvl, Body: msg})
	}
}

func BenchmarkRecordPtr(b *testing.B) {
	for i := 0; i < b.N; i++ {
		emitRecordPtr(&Record{Timestamp: t, Severity: lvl, Body: msg})
	}
}

func BenchmarkRecordOption(b *testing.B) {
	t := time.Now()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		emitWithOptions(WithTimestamp(t), WithSeverity(lvl), WithBody(msg))
	}
}
```

results:

```
enchmarkRecordValue-16         1000000000              21.79 ns/op            0 B/op          0 allocs/op
BenchmarkRecordPtr-16           1000000000              11.42 ns/op            0 B/op          0 allocs/op
BenchmarkRecordOption-16        370627761              210.3 ns/op           488 B/op          3 allocs/op
```

So the options increases allocs, but the pass via value vs pointer is still unclear.

Reply:
I made more benchmarks. So far everything confirm that using options would decrease the performance.

The value vs pointer looks similar. I would lean towards passing by value to just decrease to possibility of heap allocation.

```
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
BenchmarkRecord-16                       9944025               121.0 ns/op             0 B/op          0 allocs/op
BenchmarkRecordPtr-16                    8798300               136.9 ns/op             0 B/op          0 allocs/op
BenchmarkRecordAttrsAsField-16           8125606               144.2 ns/op           320 B/op          1 allocs/op
BenchmarkRecordPtrAttrsAsField-16        8457049               150.8 ns/op           320 B/op          1 allocs/op
BenchmarkRecordOption-16                 3001706               401.9 ns/op           832 B/op          5 allocs/op
```

```go
package main

import (
	"slices"
	"testing"
	"time"

	"go.opentelemetry.io/otel/attribute"
)

var (
	t          = time.Now()
	msg        = "log message"
	lvl        = SeverityInfo
	testFloat  = 1.2345
	testString = "7e3b3b2aaeff56a7108fe11e154200dd/7819479873059528190"
	testInt    = 32768
	testBool   = true
)

func BenchmarkRecord(b *testing.B) {
	for i := 0; i < b.N; i++ {
		r := Record{Timestamp: t, Severity: lvl, Body: msg}
		r.AddAttrs(
			attribute.String("string", testString),
			attribute.Float64("float", testFloat),
			attribute.Int("int", testInt),
			attribute.Bool("bool", testBool),
			attribute.String("string2", testString),
		)
		emitRecord(b, r)
	}
}

func BenchmarkRecordPtr(b *testing.B) {
	for i := 0; i < b.N; i++ {
		r := Record{Timestamp: t, Severity: lvl, Body: msg}
		r.AddAttrs(
			attribute.String("string", testString),
			attribute.Float64("float", testFloat),
			attribute.Int("int", testInt),
			attribute.Bool("bool", testBool),
			attribute.String("string2", testString),
		)
		emitRecordPtr(b, &r)
	}
}

func BenchmarkRecordAttrsAsField(b *testing.B) {
	for i := 0; i < b.N; i++ {
		r := RecordAttrsAsField{Timestamp: t, Severity: lvl, Body: msg, Attributes: []attribute.KeyValue{
			attribute.String("string", testString),
			attribute.Float64("float", testFloat),
			attribute.Int("int", testInt),
			attribute.Bool("bool", testBool),
			attribute.String("string2", testString)},
		}
		emitRecordAttrsAsField(b, r)
	}
}

func BenchmarkRecordPtrAttrsAsField(b *testing.B) {
	for i := 0; i < b.N; i++ {
		r := RecordAttrsAsField{Timestamp: t, Severity: lvl, Body: msg, Attributes: []attribute.KeyValue{
			attribute.String("string", testString),
			attribute.Float64("float", testFloat),
			attribute.Int("int", testInt),
			attribute.Bool("bool", testBool),
			attribute.String("string2", testString)},
		}
		emitRecordPtrAttrsAsField(b, &r)
	}
}

func BenchmarkRecordOption(b *testing.B) {
	for i := 0; i < b.N; i++ {
		emitWithOptions(WithTimestamp(t), WithSeverity(lvl), WithBody(msg), WithAttributes(
			attribute.String("string", testString),
			attribute.Float64("float", testFloat),
			attribute.Int("int", testInt),
			attribute.Bool("bool", testBool),
			attribute.String("string2", testString),
		))
	}
}

//go:noinline
func emitRecordAttrsAsField(b *testing.B, r RecordAttrsAsField) {
	if r.Body != msg {
		b.Errorf("r.Body = %s; want %s", r.Body, msg)
	}

	var found bool
	for _, kv := range r.Attributes {
		if kv.Key == "int" {
			found = true
			break
		}
	}
	if !found {
		b.Errorf("attribute int not found")
	}
}

//go:noinline
func emitRecordPtrAttrsAsField(b *testing.B, r *RecordAttrsAsField) {
	if r.Body != msg {
		b.Errorf("r.Body = %s; want %s", r.Body, msg)
	}

	var found bool
	for _, kv := range r.Attributes {
		if kv.Key == "int" {
			found = true
			break
		}
	}
	if !found {
		b.Errorf("attribute int not found")
	}
}

//go:noinline
func emitRecordPtr(b *testing.B, r *Record) {
	if r.Body != msg {
		b.Errorf("r.Body = %s; want %s", r.Body, msg)
	}

	var found bool
	r.Attrs(func(kv attribute.KeyValue) bool {
		if kv.Key == "int" {
			found = true
			return false
		}
		return true
	})
	if !found {
		b.Errorf("attribute int not found")
	}
}

//go:noinline
func emitRecord(b *testing.B, r Record) {
	if r.Body != msg {
		b.Errorf("r.Body = %s; want %s", r.Body, msg)
	}

	var found bool
	r.Attrs(func(kv attribute.KeyValue) bool {
		if kv.Key != "int" {
			return true
		}
		found = true
		return false
	})
	if !found {
		b.Errorf("attribute int not found")
	}
}

//go:noinline
func emitWithOptions(options ...RecordOption) {
	r := Record{}
	for i := 0; i < len(options); i++ {
		options[i].apply(&r)
	}
}

type RecordOption interface {
	apply(r *Record)
}

func WithTimestamp(t time.Time) RecordOption {
	return withTimestamp{t}
}

type withTimestamp struct {
	t time.Time
}

func (w withTimestamp) apply(r *Record) {
	r.Timestamp = w.t
}

func WithSeverity(s Severity) RecordOption {
	return withSeverity{s}
}

type withSeverity struct {
	s Severity
}

func (w withSeverity) apply(r *Record) {
	r.Severity = w.s
}

func WithBody(b string) RecordOption {
	return withBody{b}
}

type withBody struct {
	b string
}

func (w withBody) apply(r *Record) {
	r.Body = w.b
}

func WithAttributes(a ...attribute.KeyValue) RecordOption {
	return withAttributes{a}
}

type withAttributes struct {
	a []attribute.KeyValue
}

func (w withAttributes) apply(r *Record) {
	r.AddAttrs(w.a...)
}

type Record struct {
	Timestamp         time.Time
	ObservedTimestamp time.Time
	Severity          Severity
	SeverityText      string
	Body              string

	// The fields below are for optimizing the implementation of
	// Attributes and AddAttributes.

	// Allocation optimization: an inline array sized to hold
	// the majority of log calls (based on examination of open-source
	// code). It holds the start of the list of attributes.
	front [nAttrsInline]attribute.KeyValue

	// The number of attributes in front.
	nFront int

	// The list of attributes except for those in front.
	// Invariants:
	//   - len(back) > 0 iff nFront == len(front)
	//   - Unused array elements are zero. Used to detect mistakes.
	back []attribute.KeyValue
}

const nAttrsInline = 5

type Severity int

const (
	SeverityUndefined Severity = iota
	SeverityTrace
	SeverityTrace2
	SeverityTrace3
	SeverityTrace4
	SeverityDebug
	SeverityDebug2
	SeverityDebug3
	SeverityDebug4
	SeverityInfo
	SeverityInfo2
	SeverityInfo3
	SeverityInfo4
	SeverityWarn
	SeverityWarn2
	SeverityWarn3
	SeverityWarn4
	SeverityError
	SeverityError2
	SeverityError3
	SeverityError4
	SeverityFatal
	SeverityFatal2
	SeverityFatal3
	SeverityFatal4
)

// Attrs calls f on each attribute.KeyValue in the [Record].
// Iteration stops if f returns false.
func (r Record) Attrs(f func(attribute.KeyValue) bool) {
	for i := 0; i < r.nFront; i++ {
		if !f(r.front[i]) {
			return
		}
	}
	for _, a := range r.back {
		if !f(a) {
			return
		}
	}
}

// AddAttrs appends the given Attrs to the [Record]'s list of Attrs.
// It omits invalid attributes.
func (r *Record) AddAttrs(attrs ...attribute.KeyValue) {
	var i int
	for i = 0; i < len(attrs) && r.nFront < len(r.front); i++ {
		a := attrs[i]
		if !a.Valid() {
			continue
		}
		r.front[r.nFront] = a
		r.nFront++
	}
	// Check if a copy was modified by slicing past the end
	// and seeing if the Attr there is non-zero.
	if cap(r.back) > len(r.back) {
		end := r.back[:len(r.back)+1][len(r.back)]
		if end.Valid() {
			// Don't panic; copy and muddle through.
			r.back = slices.Clip(r.back)
			//r.back = append(r.back, String("!BUG", "AddAttrs unsafely called on copy of Record made without using Record.Clone"))
		}
	}
	ne := countInvalidAttrs(attrs[i:])
	r.back = slices.Grow(r.back, len(attrs[i:])-ne)
	for _, a := range attrs[i:] {
		if a.Valid() {
			r.back = append(r.back, a)
		}
	}
}

// countInvalidAttrs returns the number of invalid attributes.
func countInvalidAttrs(as []attribute.KeyValue) int {
	n := 0
	for _, a := range as {
		if !a.Valid() {
			n++
		}
	}
	return n
}

type RecordAttrsAsField struct {
	Timestamp         time.Time
	ObservedTimestamp time.Time
	Severity          Severity
	SeverityText      string
	Body              string
	Attributes        []attribute.KeyValue
}
```

Reply:
I did my best to describe it in a concise way. I added a hyperlink to this discussion.

https://github.com/open-telemetry/opentelemetry-go/pull/4725/commits/f288cf21afb42005de9ac1b963018c95bd9f659e

Reply:
Thanks for the benchmarks. Looks like in the pass-by-pointer approach the compiler is smart enough to realize that the Record does not escape and does not allocate it on heap, leading to pretty good performance. This may change if the code becomes more complex when the actual full implementation is done, the compiler may not be able to prove it does not escape.

In any case value vs pointer can be also decided later, since it is a relatively easy change.